theano/tensor/signal/downsample.py

```diff
@@ -518,6 +600,15 @@ def __init__(self, ds, ignore_border, st=None, padding=(0, 0), mode='max'):
                 " 'average_inc_pad' and 'average_exc_pad'. Got %s" % mode)
         self.mode = mode
 
+    def infer_shape(self, node, in_shapes):
+        return [in_shapes[0]]
+
+
+class MaxPoolGrad(PoolGrad):
```

Comment:
If I understood correctly, we would keep a `DownsampleFactorMaxGrad` with all the properties it currently has, including `mode`, but not necessarily C code, infer_shape, etc.
Then, we add an optimization to replace `DownsampleFactorMaxGrad{..., mode='max'}` by `MaxPoolGrad` and `DownsampleFactorMaxGrad{..., mode='sum/average...'}` by `AveragePoolGrad`.

There are different options:
- we could use the base `PoolGrad` class for that
- we could use `MaxPoolGrad` as @nouiz suggested, since it already has the right number of inputs
- or we could use a new, empty class (maybe inheriting from `PoolGrad`?).
I'm not sure which would be best.